nodl_python/nodl/types/_types.py

```diff
@@ -0,0 +1,169 @@
+            isinstance(other, type(self)) or isinstance(self, type(other))
+        ) and self.__dict__ == other.__dict__
+
+
+class Action(NoDLInterface):
+    """Data structure for action entries in NoDL."""
+
+    def __init__(
+        self,
+        *,
+        name: str,
+        action_type: str,
+        server: bool = False,
+        client: bool = False,
+        qos: QoSProfile = QoSPresetProfiles.ACTION_STATUS_DEFAULT.value
+    ) -> None:
+        super().__init__(name=name, value_type=action_type)
+        self.server = server
+        self.client = client
+
+        self.qos = qos
+
+
+class Parameter(NoDLInterface):
+    """Data structure for parameter entries in NoDL."""
+
+    def __init__(self, *, name: str, parameter_type: str) -> None:
+        super().__init__(name=name, value_type=parameter_type)
+
+
+class Service(NoDLInterface):
+    """Data structure for service entries in NoDL."""
+
+    def __init__(
+        self,
+        *,
+        name: str,
+        service_type: str,
+        server: bool = False,
+        client: bool = False,
+        qos: QoSProfile = QoSPresetProfiles.SERVICES_DEFAULT.value
+    ) -> None:
+        super().__init__(name=name, value_type=service_type)
+        self.server = server
+        self.client = client
+
+        self.qos = qos
+
+
+class Topic(NoDLInterface):
+    """Data structure for topic entries in NoDL."""
+
+    def __init__(
+        self,
+        *,
+        name: str,
+        message_type: str,
+        publisher: bool = False,
+        subscription: bool = False,
+        qos: QoSProfile = QoSPresetProfiles.SYSTEM_DEFAULT.value
+    ) -> None:
+        super().__init__(name=name, value_type=message_type)
+        self.publisher = publisher
+        self.subscription = subscription
+
+        self.qos = qos
+
+
+class Node(NoDLData):
+    """Data structure containing all interfaces a node exposes."""
+
+    def __init__(
+        self,
+        *,
+        name: str,
+        actions: List[Action] = [],
+        parameters: List[Parameter] = [],
+        services: List[Service] = [],
+        topics: List[Topic] = []
+    ) -> None:
+        self.name = name
+
+        self.actions = {action.name: action for action in actions}
+        self.parameters = {parameter.name: parameter for parameter in parameters}
+        self.services = {service.name: service for service in services}
+        self.topics = {topic.name: topic for topic in topics}
+
+    @property
+    def _as_dict(self):
+        return {
+            'name': self.name,
+            'actions': [action._as_dict for action in self.actions.values()],
+            'parameters': [parameter._as_dict for parameter in self.parameters.values()],
+            'services': [service._as_dict for service in self.services.values()],
+            'topics': [topic._as_dict for topic in self.topics.values()],
+        }
+
+    def __iadd__(self, other: 'Node') -> 'Node':
+        """
+        Attempt to merge two Nodes.
```

Comment:
As we discussed during standup a while back, we should not need this functionality. If the same node is specified twice, that should simply be an error.

Reply:
Removed in d785193